### ISSUE TITLE
Best Way to Get Started?

### ISSUE THREAD

**Issue by SiliconMage** [GregRos/IEMod.pw#72]
Hey all.

I'd like to get started on getting IE Mod to work with the latest patch of Pillars of Eternity.  I used to work on it in the past, but a lot has changed, so I have some questions:

1.) How does the IEMod.pw.dll get built?  Where are the mods specifically getting injected into the DLL?

2.) Do I still need to decompile the original game assembly?

3.) I noticed the PrepareEnvironment project doesn't seem to do anything or compile -- do we no longer need it?

Thanks!  I'll keep trying to hack away at this in the mean time.

**Comment by Irithyl** [comment]
Hello! I'm a IE Mod user, and just want to say a big thank you for helping with the mod! I've been stalling my play through white march II due to ie mod not being ready for it :( So a huge thank you, I really hope you can manage to make it work properly. Also people are using the nexus ie mod forum, so you might find there someone that can be of help. (http://www.nexusmods.com/pillarsofeternity/mods/1/?)

**Comment by oddrheia** [comment]
heya, I'm afraid I'm in the same boat as Irithyl here, can offer nothing constructive to help, just pass on some words of encouragement.

Maybe can get some response from the obsidian forums thread? They have an Iemod dedicated thread there: https://forums.obsidian.net/topic/81063-iemod-fork-new-modding-framework/

**Comment by Alakabon** [comment]
Hello !

I'd love to help out but can't figure out how to get started either. The project opens fine in VS but from there it seems the instructions are a bit off. The PrepareEnvironment is missing the paths files and won't run on its own. 

The patchworks system seems to work just fine but it requires the new DLLs but I can't figure out how to build them.

If anyone could give me a nudge in the right direction I'd love to contribute to this.

**Comment by tjayharvey** [comment]
Greg added a launcher since the last time I worked on this. The goal was to let people patch directly into their game files so that we don't have to distribute different binaries for Mac/Linux/Win. I'll try and track down exactly how the new front end works to try to sort out what exactly needs to be done.

**Comment by tjayharvey** [comment]
PrepareEnvironment is gone from my version. My understand is it's no longer needed for the project. Here's roughly what I think needs to happen: 
1) You need to replace the Assembly-CSharp.dll file in PoEReferences with the updated one for the patch, with all members made public. The easiest way to do this is to run the "OpenAssemblyCreator" app (in the Patchwork folder), specifying the version in your gamedir/PillarsOfEternity_Data/Managed as the source (first argument) and the version in PoEReferences as the destination.
2) This will cause IEMod itself to stop compiling, because it depended on classes/members that are no longer present in the updated source DLL. All of these compile errors have to be fixed. This usually has to be done by a) Determining whether the code in IEMod is simply trying to copy what was in the original body (done if we needed to inject code mid-member) or b) if IEMod is trying to actually inject some functionality there. This generally requires being able to look at the original source in a disassembler (I have a copy from 2.0.1 at least, so I can at least compare somewhat). If it's problem a, that's usually easily fixed by replacing it with whatever the current version has in it's body instead. If it's problem b, it requires finding a new way to do whatever IEMod wanted to do. This can take a while, so I usually start by just disabling that mod to get the rest to compile (comment out the mod file, typically).  I currently see about 16 compile errors, which isn't terribly bad.

3) Once the application is compiling again, you want to set PatchworkLauncher as the startup project and run it. It will ask you for the game directory (it doesn't save between runs, unfortunately - perhaps someone can add this?), then require you to select the mods you want to apply (by clicking "Active Mods, and adding IEMod - this also doesn't save between runs). After that you can click Test Run and see if it works successfully. It may not - not all problems show up as compile errors. These are often the trickiest to debug unfortunately. If you CAN get it to apply the mod, you can usually run Pillars with the modded DLL and start testing. Unfortunately the mod is a bit sprawling, and some of the features are quite lengthy to test. Just because we can start the app with the modded DLL doesn't mean the mod features still work - they frequently break in small ways.

I like the idea of Patchwork - it's super powerful and you can probably make PoE do literally anything you wanted it to - but it's really vulnerable to refactors in the source code, which is why the mod has a high burnout rate on maintainers.

I'll try to chip in where possible, but work is VERY busy and I have limited time available.

**Comment by tjayharvey** [comment]
Excluding from solution the following mods gets rid of most of the compile problems:
FastSneak
NoEngagement
RecoveryRate
Targeting

Of course, this disables the appropriate mod, but it's useful for testing until someone can fix them.
That left 2 problems with UICustomization (can't be disabled, other things depend on it) and one with CombatLooting. The one with CombatLooting was easily fixed (it was well commented, and was obviously a case 2a above, so I just replaced the code other than the section added by the mod with the current version and the problem went away. I was also able to fix one issue with UICustomization and get the other to the point where it would compile, if not function.

I was then able to compile the mod, test patching it, and launch PoE with the mods. It did launch successfully and I confirmed the Cipher Focus mod still works (single easiest mod to test, since it just requires loading a save game with a Cipher). This suggests the overall system is sound and will work once the mods are adjusted. 

I will try to fix FastSneak this weekend (I've fixed it several times before when it broke, so I'm probably most familiar with it). The others I have done little maintenance on, and particularly Targeting is a pain and a half to test since it involves getting your characters charmed and casting AE spells on a mix of your characters, your charmed characters and enemies (oh, and enemies you charm).

**Comment by tjayharvey** [comment]
Actually, apparently the game directory and selected mods DO save after the first time you successfully patch. But that's still kind of annoying :).

**Comment by SiliconMage** [comment]
I haven't had a problem compiling, but how can you exclude mods from the build?  It used to be that when you applied the mod, you could tell exactly which mod was breaking, but now I seem to be a little lost on how to tell.  :)

Does rebuilding the IEMod.pw solution remake the .dll needed for the Launcher to apply to the game assembly dll?

**Comment by llerelol** [comment]
Hey people, just wanted to add my 2 cents.

First of all tjayharvey thank you for your work, you're our last hope :)
Is there a chance you could make separate FastStealth mod, like standalone? I'm sure there are a lot of people like me who only use that one out of all features.

And second, since there's a launcher now and you can activate mods one by one wouldn't it be better to separate all the modules IE Mod has into smaller addons. It would be much easier to maintain and we can always choose to ignore those that break after every update.

**Comment by tjayharvey** [comment]
SiliconMage: 
Yes it does. To exclude mods from the build, you just exclude them from the IEMod project. IEMod gets compiled into a DLL that is used by Patchwork (through reflection and attributes) to build a manifest of changes to apply to the source DLL, which Patchwork then applies (Patchwork is a generic .NET DLL patcher, and isn't specific to IEMod or even Pillars of Eternity).

If you aren't getting compile errors for IEMod, then you haven't updated the IEMod project's reference to Assembly-CSharp.dll to the latest version. It does look like it saves in the repository, so I can check mine in and push to the main repository if that's easier.

**Comment by tjayharvey** [comment]
Ilerelol - That was kind of the intention for the future of IEMod. There's three issues that need to be dealt with before we really can split them out:

1) Better error messages for patch failure. Right now you get a very ambiguous error that doesn't really help you track down the problem much. This would lead to users being unsure of which mod dll was causing the issue, and the maintainer of each individual piece with very limited understanding of what went wrong.

2) Conflict detection - if two mods both want to modify the body of the same method, this isn't going to work as expected (the second one in load order will probably win, but it might even fail to patch it). Of course, IEMod as a unified whole suffers from this too, but it's way easier to see conflicts when they are in the same solution.

3) Pluggable mod setting UI. Most mods (including Fast Sneak!) have multiple options that can be selected. Asking the user to juggle DLLs and selecting only the ones for the options they want when they run the game is awkward. If we had some kind of base mod that handled mod options in a generic way that other mods could then plug into, it would be much easier to have something like Fast Sneak in its own DLL that you could apply or not, and then select the mod settings within the game itself.

I've actually already fixed Fast Sneak, so you should be able to use it again fairly soon..

**Comment by llerelol** [comment]
yeah, i've seen you fixed it thats why i asked.
my point is i just want standalone Fast Sneak since i dont want other modules to break stuff.
maybe if you can point me into right direction i can make it myself.

if i get it right the only way you can patch assembly is to edit CIL code. and im totaly not familiar with it. im also aware of cecil lib, but thats a lot of work also. maybe there's another way?

**Comment by Alakabon** [comment]
I've got lots of time this week-end/ next week if you need help with the testing.

**Comment by tjayharvey** [comment]
Ilerlol: You can probably make it from the existing IEMod.by removing all the other mods and renaming things appropriately. IEMod is essentially declarative - it consists of various classes describing what the code in the target assembly should be changed to. You can limit it to just the fast sneak changes and alter the manifest that Patchwork uses to tell what it is and distribute it separately (Patchwork will never break in an PoE patch, so you don't have to worry about that).

I will say that Fast Sneak is one of the most frequently broken mods (it's in the "main game loop" for party members), so I'm not sure the effort will help much in terms of reducing churn between patches.

**Comment by tjayharvey** [comment]
New version is up on the Nexus site if you want to test. I still need to merge back to the main repo (for those of you playing with the code).

**Comment by SiliconMage** [comment]
tjayharvey,

Thanks for your advice!  I have the solution building properly now, and I can apply the IEMod.pw.dll to my assembly dll successfully!  :+1: 

I'll work on testing as well, and I may add a few tweaks I have in mind.  I'll try to update my branch specifically and make pull requests into this one with any updates I have.  Thanks again!

**Comment by tjayharvey** [comment]
Do you have merge permissions to this fork? Are you working on your own fork? Also - if you are working on something in particular it's good to start an Issue for that particular feature (like I did with Per Encounter Spells). You can then assign it to yourself (or I can if you don't have permissions).

**Comment by SiliconMage** [comment]
I don't believe I have permissions to merge to this fork.  I was working in my own fork, sorta.  I haven't updated it in the longest time.  

As far as issues, I'm going to look into Disable Friendly Fire and Target Turned Enemies to make sure that either of them aren't causing crashes.

Unless there is something else that is more important.  I was looking at the IEMod thread on the Nexus site, and some people are still complaining about crashes.

I had to use Telerik's JustDecompile to get most of the code decompiled -- ILSpy kept crashing.

**Comment by tjayharvey** [comment]
Yeah, I use JustDecompile as well. I'd say the combat crashes are by far the biggest priority atm. If you come up with a fix, you can post it to your fork or wherever and send a pull request - I have permission to merge into the master repo (merged all my fixes in already).

**Comment by SiliconMage** [comment]
I updated the code for Disable Friendly Fire.  I haven't experienced any of the crashes mentioned since updating it.  I created a pull request for it.

I also added another mod for hiding Anticlass spells from the ability bar.

**Comment by Alakabon** [comment]
Hi, I could use a bit of help if you guys have a minute...

Essentially, I'm adding a new console command to load stats from an xml file, trying to work around the new system who resets them after every reload. I have 2 problems:

1) When I type the command in the console in the game, it won't let me... "Activate" it, nothing happens when I press enter. I'm 99% certain it recognizes the command since I'm not getting the command not found error message I'm getting with other random text and if I add nonsense arguments it tells me it's not the right number of arguments... 
Does this just mean its executing but I'm doing something wrong or do I have to write additionnal code somewhere to get it to launch ? The only file I modified is Mod_CommandLine.cs.

2) I'm not too sure how the Pathing works when I'm asking it to read the file. Does Application.dataPath give me the PillarsOfEternity_Data folder, or do I end up somewhere else?

**Comment by tjayharvey** [comment]
I'd recommend adding logging calls to see what's going on. It's possible the code is running into an error. You can look at Alter Resources Mod for an example of reading from a configuration file. It looks like Application.dataPath should give you PillarsOfEternity_Data folder. Here's what my call looks like to load ability data from XML:

`            AbilityActionData result = new AbilityActionData();

```
        string abilityDataPath = Path.Combine(Path.Combine(Application.dataPath, "Managed/iemod/data"), "abilitydata.xml");
        IEDebug.Log($"Looking for file {abilityDataPath}");
        if (File.Exists(abilityDataPath))
        {
            IEDebug.Log("Found abilitydata.xml");
            try
            {
                XmlSerializer serializer = new XmlSerializer(typeof(Serialization.AbilityActionData));
                Serialization.AbilityActionData serializedData;
                using (var openRead = File.OpenRead(abilityDataPath))
                {
                    serializedData = (Serialization.AbilityActionData)serializer.Deserialize(openRead);
                }

                result = new AbilityActionData(serializedData);

            }
            catch (Exception ex)
            {
                IEDebug.Log($"Couldn't process ability data: {ex.ToString()}");
            }
        }

        return result;`
```

**Comment by Alakabon** [comment]
Thank you !

I got it to work, turns out I used .Value instead of .InnerText when parsing, rookie mistake. I took the IEDebug logs back out since the class said it might not be great to use it on release.

It's not 100% yet, as it is now I'm not accounting for the racial bonuses, but otherwise its functionnal. It seems quick enough but I guess it could slowdown a tad depending on the amount characters in the area.

Since I only changed the one file do you prefer if I paste the code in here ?

**Comment by Irithyl** [comment]
Hey! How is the mod going? When do you think it will be ready? :)

**Comment by SiliconMage** [comment]
Is there any way I could get admin access to the branch?  I would like to release an updated version of the mod, and incorporate the latest changes into it.

**Comment by tjayharvey** [comment]
I'm not sure I have permission to do that, but I will see what I can do. In the meantime I can pull a release if you like.

**Comment by tjayharvey** [comment]
Just resolved conflicts, built, and released 5.2.1.

I also fixed the Version Number UI Mod so you don't have to update the base game version any more (that was one of the conflicts, actually).

**Comment by Alakabon** [comment]
There will be a bit of a problem if people try using the autoload stats without the xml files in the Managed/iemod/customStats... Perhaps include them in the zip so they have the xml structure to follow?

**Comment by tjayharvey** [comment]
Where are the XML files located in the solution?

**Comment by Alakabon** [comment]
IEMod/iemod/customStats/

I put a readme in there so people understand the limitation of the current system as well.

**Comment by tjayharvey** [comment]
I added the two XML files and the Readme to the DLL. (I also grabbed the wrong DLL the first time. Oops!) Probably need to think about making a "Publish" app that packages up the Zip for upload to Nexus.

**Comment by Alakabon** [comment]
I hate to be that guy but... I only see the DLL in the downloaded zip ?

**Comment by tjayharvey** [comment]
Try now. I seem to fail at Nexus today.

**Comment by Alakabon** [comment]
All there ! I think the readme includes the intended path, I reckon people are going to figure out where they go. Otherwise I'll try and spread the knowledge around on reddit/nexus if need be.

**Comment by tjayharvey** [comment]
This thread has gotten super-long - I've started another one for the next release (and pulled out the "people need access" into its own issue). 

Please close any issues fixed in the 5.2.1 release.
